src/gluonnlp/sequence_sampler.py

```diff
@@ -717,11 +717,20 @@ def forward(self, samples, valid_length, outputs, scores, step, beam_alive_mask,
         probs = mx.npx.softmax(outputs / self._temperature)
 
         if self._sampling_topp > 0:
-            probs = mx.np.where(
-                probs > self._sampling_topp,
-                probs,
-                mx.np.zeros_like(probs)
-            )
+             sorted_probs, sorted_indices = mx.npx.topk(probs, axis=2, k=-1, ret_typ='both', is_ascend=False)
+             cumsum_probs = mx.np.cumsum(sorted_probs, axis=2)
+             masked_probs = mx.np.where(
+                 cumsum_probs > self._sampling_topp,
+                 sorted_probs,
+                 mx.np.ones_like(probs)
+             )
+             # choose the borderline prob
+             p_prob = mx.np.min(masked_probs, axis=2, keepdims=True)
+             probs = mx.np.where(
+                 probs >= p_prob,
+                 probs,
+                 mx.np.zeros_like(probs)
+             )
```

Comment:
The major difference between the current implementation and the original pytorch-based implementation is that when ``sampling_topp < max(probs)``, it is not clear which probability will be picked.

The pytorch-based implementation will always choose the token that are most probable.